### PR TITLE
Tensorboard integration

### PR DESCRIPTION
**Pull request by jantic**
This is an initial commit of tensorboard functionality for fastai.  Note that you'll want to take into account that tensorboardX and tensorboard are not yet by default installed.  Jeremy said he'd be the one to add error handling code around this.

Here's what's going up currently:

* Image generation visualization (for both GAN and non GAN learners)
* Model histograms/distributions for each of the layers
* Various gradient stats
* All available metrics/losses (losses are reported as a 'metric' currently).

These are callbacks for Learners.   The usage is as follows (example directly from DeOldify):
```
    proj_id = 'Colorize'
    tboard_path = Path('data/tensorboard/' + proj_id)
    learn.callback_fns.append(partial(GANTensorboardWriter, base_dir=tboard_path, name='GanLearner'))
```

More details can be found in fast-ai dev thread on this subject.

### CODE REVIEW

**Comment by terribilissimo** [comment]
Thanks for this. I was thinking about doing it too, then I saw your PR.

File: fastai/callbacks/tensorboard.py [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+            orig_images.append(image_set.orig.px)
+            gen_images.append(image_set.gen.px)
+            real_images.append(image_set.real.px) 
+        return orig_images, gen_images, real_images  
+
+    def write(self)->None:
+        "Writes original, generated and real(target) images to Tensorboard."
+        orig_images, gen_images, real_images = self._get_image_tensors()
+        self._write_images(name='orig images', images=orig_images)
+        self._write_images(name='gen images', images=gen_images)
```
Comment (sgugger) [inline fastai/callbacks/tensorboard.py]: A little space here after the comma to align everything nicely.

File: fastai/callbacks/tensorboard.py [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+
+    def _write_images(self, iteration:int)->None:
+        "Writes model generated, original and real images to Tensorboard."
+        trainer = self.learn.gan_trainer
+        #TODO:  Switching gen_mode temporarily seems a bit hacky here.  Certainly not a good side-effect.  Is there a better way?
+        gen_mode = trainer.gen_mode
+        try:
+            trainer.switch(gen_mode=True)
+            self.img_gen_vis.write(learn=self.learn, trn_batch=self.trn_batch, val_batch=self.val_batch, iteration=iteration, tbwriter=self.tbwriter)
+        finally:
```
Comment (sgugger) [inline fastai/callbacks/tensorboard.py]: On the same line here

File: fastai/callbacks/tensorboard.py [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+        self.visual_iters = visual_iters
+        self.img_gen_vis = ImageTBWriter()
+        self.gen_stats_updated = True
+        self.crit_stats_updated = True
+
+    def _write_weight_histograms(self, iteration:int)->None:
+        "Writes model weight histograms to Tensorboard."
+        generator, critic = self.learn.gan_trainer.generator, self.learn.gan_trainer.critic
+        self.hist_writer.write(model=generator, iteration=iteration, tbwriter=self.tbwriter, name='generator')
+        self.hist_writer.write(model=critic, iteration=iteration, tbwriter=self.tbwriter, name='critic')
```
Comment (sgugger) [inline fastai/callbacks/tensorboard.py]: Vertically align with previous line to make it nicer (iteration, tbwriter and name)

File: fastai/callbacks/tensorboard.py [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+        if iteration % self.stats_iters == 0: self._write_model_stats(iteration=iteration)
+
+    def on_epoch_end(self, last_metrics:MetricsList, iteration:int, **kwargs)->None:
+        "Callback function that writes epoch end appropriate data to Tensorboard."
+        self._write_metrics(iteration=iteration, last_metrics=last_metrics)
+
+# TODO:  We're overriding almost everything here.  Seems like a good idea to question that ("is a" vs "has a")
+class GANTensorboardWriter(LearnerTensorboardWriter):
+    "Callback for GANLearners that writes to Tensorboard.  Extends LearnerTensorboardWriter and adds output image writes."
+    def __init__(self, learn:GANLearner, base_dir:Path, name:str, loss_iters:int=25, hist_iters:int=500, stats_iters:int=100, visual_iters:int=100):
```
Comment (sgugger) [inline fastai/callbacks/tensorboard.py]: Maybe those two lines need breaking for GitHub handling ;)

**Comment by sgugger** [comment]
Thanks a lot! There are just a last few cosmetic changes to make, and one import to remove, then we'll be good to merge.

**Comment by sgugger** [comment]
Just a few line breaks and no spaces between multiple assignments (a,b = bli,bla)
Thanks a lot for all your work!

**Comment by jantic** [comment]
Thanks @sgugger!

**Comment by panaali** [comment]
@jantic I have problem with the this line : [32](https://github.com/fastai/fastai/blob/c32d553d4afd32caab93db3e07a27c1048b85044/fastai/callbacks/tensorboard.py#L32) ```  self.tbwriter = SummaryWriter(log_dir=str(log_dir))```
It seems that [SummaryWriter](https://github.com/lanpa/tensorboardX/blob/7669082fb2504e5730ea8a19853de786a92bf8c7/tensorboardX/writer.py#L191)'s input for TensorboardX should be `logdir` not `log_dir`.
It gives me ```TypeError: __init__() got an unexpected keyword argument 'log_dir'``` error

**Comment by jantic** [comment]
> @jantic I have problem with the this line : [32](https://github.com/fastai/fastai/blob/c32d553d4afd32caab93db3e07a27c1048b85044/fastai/callbacks/tensorboard.py#L32) ` self.tbwriter = SummaryWriter(log_dir=str(log_dir))`
> It seems that [SummaryWriter](https://github.com/lanpa/tensorboardX/blob/7669082fb2504e5730ea8a19853de786a92bf8c7/tensorboardX/writer.py#L191)'s input for TensorboardX should be `logdir` not `log_dir`.
> It gives me `TypeError: __init__() got an unexpected keyword argument 'log_dir'` error

Facepalm.  They changed the variable name recently:

https://github.com/lanpa/tensorboardX/commit/bf8c6795779fbf225a6bbfefb5a53f9191e7440e#diff-0900f72201599ce7d24575f21ca88a85

I'll just change it to be a positional argument to make things easy and submit a pull request.  Thanks!

**Pull request merged**